staging/src/k8s.io/apiserver/pkg/storage/cacher/caching_object.go

```diff
@@ -0,0 +1,404 @@
+	o.invalidateCacheLocked()
+	o.object.SetGenerateName(name)
+}
+func (o *cachingObject) GetUID() types.UID {
+	o.lock.RLock()
+	defer o.lock.RUnlock()
+	return o.object.GetUID()
+}
+func (o *cachingObject) SetUID(uid types.UID) {
+	o.lock.Lock()
+	defer o.lock.Unlock()
+	if o.object.GetUID() == uid {
+		return
+	}
+	o.invalidateCacheLocked()
+	o.object.SetUID(uid)
+}
+func (o *cachingObject) GetResourceVersion() string {
+	o.lock.RLock()
+	defer o.lock.RUnlock()
+	return o.object.GetResourceVersion()
+}
+func (o *cachingObject) SetResourceVersion(version string) {
+	o.lock.Lock()
+	defer o.lock.Unlock()
+	if o.object.GetResourceVersion() == version {
+		return
+	}
+	o.invalidateCacheLocked()
+	o.object.SetResourceVersion(version)
+}
+func (o *cachingObject) GetGeneration() int64 {
+	o.lock.RLock()
+	defer o.lock.RUnlock()
+	return o.object.GetGeneration()
+}
+func (o *cachingObject) SetGeneration(generation int64) {
+	o.lock.Lock()
+	defer o.lock.Unlock()
+	if o.object.GetGeneration() == generation {
+		return
+	}
+	o.invalidateCacheLocked()
+	o.object.SetGeneration(generation)
+}
+func (o *cachingObject) GetSelfLink() string {
+	o.lock.RLock()
+	defer o.lock.RUnlock()
+	return o.object.GetSelfLink()
+}
+func (o *cachingObject) SetSelfLink(selfLink string) {
+	o.lock.Lock()
+	defer o.lock.Unlock()
+	if o.object.GetSelfLink() == selfLink {
+		return
+	}
+	o.invalidateCacheLocked()
+	o.object.SetSelfLink(selfLink)
+}
+func (o *cachingObject) GetCreationTimestamp() metav1.Time {
+	o.lock.RLock()
+	defer o.lock.RUnlock()
+	return o.object.GetCreationTimestamp()
+}
+func (o *cachingObject) SetCreationTimestamp(timestamp metav1.Time) {
+	o.lock.Lock()
+	defer o.lock.Unlock()
+	if o.object.GetCreationTimestamp() == timestamp {
+		return
+	}
+	o.invalidateCacheLocked()
+	o.object.SetCreationTimestamp(timestamp)
+}
+func (o *cachingObject) GetDeletionTimestamp() *metav1.Time {
+	o.lock.RLock()
+	defer o.lock.RUnlock()
+	return o.object.GetDeletionTimestamp()
+}
+func (o *cachingObject) SetDeletionTimestamp(timestamp *metav1.Time) {
+	o.lock.Lock()
+	defer o.lock.Unlock()
+	if o.object.GetDeletionTimestamp() == timestamp {
+		return
+	}
+	o.invalidateCacheLocked()
+	o.object.SetDeletionTimestamp(timestamp)
+}
+func (o *cachingObject) GetDeletionGracePeriodSeconds() *int64 {
+	o.lock.RLock()
+	defer o.lock.RUnlock()
+	return o.object.GetDeletionGracePeriodSeconds()
+}
+func (o *cachingObject) SetDeletionGracePeriodSeconds(gracePeriodSeconds *int64) {
+	o.lock.Lock()
+	defer o.lock.Unlock()
+	if o.object.GetDeletionGracePeriodSeconds() == gracePeriodSeconds {
+		return
+	}
+	o.invalidateCacheLocked()
+	o.object.SetDeletionGracePeriodSeconds(gracePeriodSeconds)
+}
+func (o *cachingObject) GetLabels() map[string]string {
+	o.lock.RLock()
+	defer o.lock.RUnlock()
+	return o.object.GetLabels()
+}
+func (o *cachingObject) SetLabels(labels map[string]string) {
+	o.lock.Lock()
+	defer o.lock.Unlock()
+	if equality.Semantic.DeepEqual(o.object.GetLabels(), labels) {
```

Comment:
check if nil vs {} produces different serializations. I think we might want to make this stronger (reflect.DeepEqual) so that *any* difference invalidates

Reply:
reflect.DeepEqual is safer, we just risk unnecessary cache invalidations (and we're not actually calling SetLabels in any known paths). same comment applies to all DeepEqual comparisons

Reply:
done